Core/Graphics/Source/SafeTimespanGuarantor.cpp

```diff
@@ -0,0 +1,70 @@
+#include "SafeTimespanGuarantor.h"
+
+namespace Babylon
+{
+    SafeTimespanGuarantor::SafeTimespanGuarantor()
+        : m_affinity{ std::this_thread::get_id() }
+        , m_lock{ m_mutex }
+    {
+    }
+
+    void SafeTimespanGuarantor::BeginSafeTimespan()
+    {
+        if (!m_affinity.check())
+        {
+            throw std::runtime_error{ "BeginSafeTimespan must be called from the thread on which the SafeTimespanGuarantor was constructed." };
+        }
+
+        if (!m_lock.owns_lock())
+        {
+            throw std::runtime_error{ "EndSafeTimespan must be called before BeginSafeTimespan can be called again." };
+        }
+
+        // First unlock the underlying mutex, which allows calls to GetSafetyGuarentee to acquire a SafetyGuarentee.
+        m_lock.unlock();
+
+        // Then yield to ensure calls to GetSafetyGuarentee get a chance to lock on the mutex before EndSafeTimeSpan takes the lock (e.g. prevent starvation).
+        std::this_thread::yield();
```

Comment:
What's a braking mutex?

Reply:
The problem I believe this yield is trying to address -- that it's possible for a single thread looping over a mutex to unlock that mutex, do some work, then relock it without giving other threads that wanted to do so a chance to acquire their own locks -- was something Gary and I identified in an earlier change around the same time we were discussing what became the `SafeTimespanGuarantor`. The braking mutex was a special mutex intended to not have that problem; it was developed alongside the `SafeTimespanGuarantor` and was the first mutex used within it. It was taken out because the problem it was intended to solve stopped showing up, so we never really tried it out much, but it's still there in the history. If you look back through this PR, it's around in the first few of my commits.